### ISSUE TITLE
PG::InsufficientPrivilege: ERROR:  permission denied to create extension "hstore"

### ISSUE THREAD

**Issue by wixyvir** [bitnami/bitnami-docker-discourse#105]
**Description**

I am using "latest" docker images, when I instantiate discourse a first time, it works fine. But if I drop the discourse container, and re-instantiate one with the same volumes (on `/bitnami`), then it won't properly start because entrypoint script is trying to migrate database again, and is trying to create PG extensions again, as a non super-user.

**Steps to reproduce the issue:**

1. docker-compose up # with bind volumes on /bitnami
2. Set up discourse, log-in, configure...
3. Drop the discourse service container
4. Re-instantiate a discourse service container # with same volumes as the one in 1.

**Describe the results you received:**

It does not start and produce the following error

```
Welcome to the Bitnami discourse container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-discourse
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-discourse/issues

nami    INFO  Initializing postgresql-client
postgre INFO  Trying to connect to PostgreSQL server
postgre INFO  Found PostgreSQL server listening at postgresql:5432
postgre INFO  PostgreSQL server listening and working at postgresql:5432
nami    INFO  postgresql-client successfully initialized
nami    INFO  Initializing discourse
discour INFO  Patching Discourse...
postgre INFO  Trying to connect to PostgreSQL server
postgre INFO  Found PostgreSQL server listening at postgresql:5432
postgre INFO  PostgreSQL server listening and working at postgresql:5432
discour INFO  Preparing Discourse environment...
discour INFO  Configuring the Discourse database files...
discour INFO  Configuring Redis...
discour INFO  Setting up SMTP in Discourse...
discour INFO  Migrating database...
Error executing 'postInstallation': rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::InsufficientPrivilege: ERROR:  permission denied to create extension "hstore"
HINT:  Must be superuser to create this extension.
: CREATE EXTENSION IF NOT EXISTS hstore
/opt/bitnami/discourse/vendor/bundle/ruby/2.5.0/gems/rack-mini-profiler-1.0.2/lib/patches/db/pg.rb:110:in `async_exec'
[...]

Caused by:
ActiveRecord::StatementInvalid: PG::InsufficientPrivilege: ERROR:  permission denied to create extension "hstore"
HINT:  Must be superuser to create this extension.
: CREATE EXTENSION IF NOT EXISTS hstore
/opt/bitnami/discourse/vendor/bundle/ruby/2.5.0/gems/rack-mini-profiler-1.0.2/lib/patches/db/pg.rb:110:in `async_exec'
[...]

Caused by:
PG::InsufficientPrivilege: ERROR:  permission denied to create extension "hstore"
HINT:  Must be superuser to create this extension.
/opt/bitnami/discourse/vendor/bundle/ruby/2.5.0/gems/rack-mini-profiler-1.0.2/lib/patches/db/pg.rb:110:in `async_exec'
[...]
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

**Describe the results you expected:**

Discourse should be available

**Additional information you deem important (e.g. issue happens only occasionally):**

Also is there a way to know what this is doing https://github.com/bitnami/bitnami-docker-discourse/blob/master/2/debian-9/Dockerfile#L5-L11 ? Where are those packages sources ?

**Version**

- Output of `docker version`:

```
Client:
 Version:           18.09.8
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        0dd43dd87f
 Built:             Wed Jul 17 17:40:56 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       4d60db4
  Built:            Wed Nov  7 00:19:08 2018
  OS/Arch:          linux/amd64
  Experimental:     true
```

- Output of `docker info`:

```
Containers: 28
 Running: 26
 Paused: 0
 Stopped: 2
Images: 31
Server Version: 18.09.0
Storage Driver: devicemapper
 Pool Name: centos_tech-thinpool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Udev Sync Supported: true
 Data Space Used: 24.19GB
 Data Space Total: 42.95GB
 Data Space Available: 18.76GB
 Metadata Space Used: 24.07MB
 Metadata Space Total: 2.147GB
 Metadata Space Available: 2.123GB
 Thin Pool Minimum Free Space: 4.295GB
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.149-RHEL7 (2018-07-20)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: active
 NodeID: xv2y1ldk3u1ppb19gl4cf58jp
 Is Manager: true
 ClusterID: r4c2zh4pkmmk8u5tp2pbj52gf
 Managers: 1
 Nodes: 1
 Default Address Pool: 10.0.0.0/8  
 SubnetSize: 24
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.0.0.75
 Manager Addresses:
  10.0.0.75:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: c4446665cb9c30056f4998ed953e6d4ff22c7c39
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.20.4-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.776GiB
Name: srv-tool0
ID: FS73:MACJ:V3CW:72SP:PEI2:7TNZ:CIF6:XCEU:EQ26:TNGP:R6WR:BVCQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 env=tools
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: bridge-nf-call-ip6tables is disabled
```

Thank you for your help !

**Comment by tompizmor** [comment]
Hi @wixyvir,

I tried with this docker-compose file and it worked fine after running `docker-compose up`, `docker-compose down` and `docker-compose up` again.

```
version: '2'
services:
  postgresql:
    image: 'bitnami/postgresql:11'
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
    volumes:
      - 'postgresql_data:/bitnami/postgresql'
  redis:
    image: 'bitnami/redis:5.0'
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
    volumes:
      - 'redis_data:/bitnami'
  discourse:
    image: 'bitnami/discourse:2'
    ports:
      - '80:3000'
    depends_on:
      - postgresql
      - redis
    volumes:
      - 'discourse_data:/bitnami'
    environment:
      - POSTGRESQL_HOST=postgresql
      - POSTGRESQL_ROOT_USER=postgres
      - POSTGRESQL_CLIENT_CREATE_DATABASE_NAME=bitnami_application
      - POSTGRESQL_CLIENT_CREATE_DATABASE_USERNAME=bn_discourse
      - POSTGRESQL_CLIENT_CREATE_DATABASE_PASSWORD=bitnami1
      - DISCOURSE_POSTGRESQL_NAME=bitnami_application
      - DISCOURSE_POSTGRESQL_USERNAME=bn_discourse
      - DISCOURSE_POSTGRESQL_PASSWORD=bitnami1
      - DISCOURSE_HOSTNAME=127.0.0.1
  sidekiq:
    image: 'bitnami/discourse:2'
    depends_on:
      - discourse
    volumes:
      - 'sidekiq_data:/bitnami'
    command: 'nami start --foreground discourse-sidekiq'
    environment:
      - DISCOURSE_POSTGRESQL_NAME=bitnami_application
      - DISCOURSE_POSTGRESQL_USERNAME=bn_discourse
      - DISCOURSE_POSTGRESQL_PASSWORD=bitnami1
      - DISCOURSE_HOST=discourse
      - DISCOURSE_PORT=3000
      - DISCOURSE_HOSTNAME=127.0.0.1
volumes:
  postgresql_data:
    driver: local
  redis_data:
    driver: local
  discourse_data:
    driver: local
  sidekiq_data:
    driver: local
```

> Also is there a way to know what this is doing https://github.com/bitnami/bitnami-docker-discourse/blob/master/2/debian-9/Dockerfile#L5-L11 ? Where are those packages sources ?

You can see the source files inside the container at `/root/.nami/components`.

There are some components that are installed directly in the Dockerfile while others are unpacked in the Dockerfile and initialized in the entrypoint.sh. So basically `bitnami-pkg install` = `bitnami-pkg unpack` + `bitnami-pkg initialize`

**Comment by mmacfadden** [comment]
@tompizmor I am seeing this same issue now.  In your example, when you did docker compose down, would the postgresql volume have gone away?

**Comment by mmacfadden** [comment]
One other bit of info.. looking at the output of the postgresql container I see this:
```
postgresql 19:52:52.10 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql
postgresql 19:52:52.10 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql/issues
postgresql 19:52:52.10 Send us your feedback at containers@bitnami.com
postgresql 19:52:52.11 
postgresql 19:52:52.12 INFO  ==> ** Starting PostgreSQL setup **
postgresql 19:52:52.16 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql 19:52:52.16 INFO  ==> Loading custom pre-init scripts...
postgresql 19:52:52.17 INFO  ==> Initializing PostgreSQL database...
postgresql 19:52:52.18 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql 19:52:52.20 INFO  ==> pg_hba.conf file not detected. Generating it...
postgresql 19:52:53.41 INFO  ==> Starting PostgreSQL in background...
postgresql 19:52:53.53 INFO  ==> Changing password of postgres
postgresql 19:52:53.54 INFO  ==> Configuring replication parameters
postgresql 19:52:53.55 INFO  ==> Configuring fsync
postgresql 19:52:53.56 INFO  ==> Loading custom scripts...
postgresql 19:52:53.56 INFO  ==> Enabling remote connections
postgresql 19:52:53.57 INFO  ==> Stopping PostgreSQL...
postgresql 19:52:54.58 INFO  ==> ** PostgreSQL setup finished! **

postgresql 19:52:54.62 INFO  ==> ** Starting PostgreSQL **
2019-11-11 19:52:54.630 GMT [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2019-11-11 19:52:54.630 GMT [1] LOG:  listening on IPv6 address "::", port 5432
2019-11-11 19:52:54.634 GMT [1] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2019-11-11 19:52:54.645 GMT [195] LOG:  database system was shut down at 2019-11-11 19:52:53 GMT
2019-11-11 19:52:54.650 GMT [1] LOG:  database system is ready to accept connections
2019-11-11 19:53:09.705 GMT [202] LOG:  incomplete startup packet
2019-11-11 19:53:09.709 GMT [203] LOG:  incomplete startup packet
2019-11-11 19:53:09.741 GMT [205] FATAL:  password authentication failed for user "bn_discourse"
2019-11-11 19:53:09.741 GMT [205] DETAIL:  Role "bn_discourse" does not exist.
	Connection matched pg_hba.conf line 1: "host     all             all             0.0.0.0/0               md5"
2019-11-11 19:53:13.894 GMT [213] LOG:  incomplete startup packet
2019-11-11 19:53:25.653 GMT [216] ERROR:  permission denied to create extension "hstore"
2019-11-11 19:53:25.653 GMT [216] HINT:  Must be superuser to create this extension.
2019-11-11 19:53:25.653 GMT [216] STATEMENT:  CREATE EXTENSION IF NOT EXISTS hstore
```

I noticed the `2019-11-11 19:53:09.741 GMT [205] FATAL:  password authentication failed for user "bn_discourse"` line.  I am wondering if authentication is failing and therefore discourse thinks the database does not exist.

**Comment by tompizmor** [comment]
> In your example, when you did docker compose down, would the postgresql volume have gone away?

No, because I run `docker-compose down` and not `docker-compose down -v`.

I tried to reproduce it again but I couldn't. Are you running `docker-compose up`, `docker-compose down`, and `docker-compose up` again or you are doing something different?

**Comment by mmacfadden** [comment]
@tompizmor We actually have this running in Kubernetes.  I could not seem to resolve the issue, and we didn't have much content in discourse yet, so I just blew away the installation and reinstalled in from scratch. This was a few days ago. However, we are now back in the same situation.  Basically here is what happened:

1. Deployed the bitnami versions of postgresql, redis, and discourse / sidekiq. PostgreSQL is set up with a persistent volume.
2. Since PostgreSQL was empty Discourse initialized the schema, and created an empty site.
3. We used the setup wizard to configure the discourse site.
4. Occasionally, due to maintenance or rebalancing the discourse container would be migrated to another node, causing it to stop and restart. Discourse would go down, during this time briefly, but come back up.
5. A few days later, we are now back in the same situation as I described above and Discourse will not start.

```
PG::InsufficientPrivilege: ERROR:  permission denied to create extension "hstore"
HINT:  Must be superuser to create this extension.
: CREATE EXTENSION IF NOT EXISTS hstore
```

**Comment by tompizmor** [comment]
Do you think you can share your kubernetes manifests so I can try to reproduce in a similar environment?

**Comment by mmacfadden** [comment]
@tompizmor Yes I think that is possible. It may take a couple of days.

**Comment by mmacfadden** [comment]
Working on posting my configs now.  One tthing I noticed, is that I `exec` into the postgres container and execute the following:

```
$ psql -A -t -U bn_discourse -W -d bitnami_discourse
Password: 
psql (11.5)
Type "help" for help.

bitnami_discourse=> \dt
public|ar_internal_metadata|table|bn_discourse
public|archetypes|table|bn_discourse
public|categories|table|bn_discourse
public|category_featured_threads|table|bn_discourse
public|category_featured_users|table|bn_discourse
public|forum_thread_link_clicks|table|bn_discourse
public|forum_thread_links|table|bn_discourse
public|forum_thread_users|table|bn_discourse
public|forum_threads|table|bn_discourse
public|incoming_links|table|bn_discourse
public|message_bus|table|bn_discourse
public|notifications|table|bn_discourse
public|onebox_renders|table|bn_discourse
public|post_action_types|table|bn_discourse
public|post_actions|table|bn_discourse
public|post_onebox_renders|table|bn_discourse
public|post_replies|table|bn_discourse
public|post_timings|table|bn_discourse
public|posts|table|bn_discourse
public|schema_migration_details|table|bn_discourse
public|schema_migrations|table|bn_discourse
public|site_settings|table|bn_discourse
public|twitter_user_infos|table|bn_discourse
public|uploads|table|bn_discourse
public|user_actions|table|bn_discourse
public|user_open_ids|table|bn_discourse
public|users|table|bn_discourse
public|versions|table|bn_discourse
public|views|table|bn_discourse
bitnami_discourse=> SELECT * FROM users;
bitnami_discourse=> SELECT * FROM versions;
```

It somewhat looks like the database is empty?  I assume that since I created several users they should show up in the users table?

**Comment by marcosbc** [comment]
Hi @mmacfadden, what version of the Discourse container are you using? In Discourse 2.3.7 the "versions" table is not present.

In any case, it looks like your persistent storage is not working properly. Keep in mind that it must be persisting the following directories:

- PostgreSQL: `/bitnami/postgresql` (**not** `/bitnami`, that won't work)
- Discourse: `/bitnami/discourse` (in this case `/bitnami` would work)
- Sidekiq: `/bitnami/discourse-sidekiq` (in this case `/bitnami` would work)

We've found similar issues when persisting the incorrect directory in PostgreSQL, due to `/bitnami/postgresql` being considered as a Docker volume. In those cases, when you mounted `/bitnami`, `/bitnami/postgresql` would be considered as independent and therefore its contents would be emptied after successive restarts.

**Comment by mmacfadden** [comment]
@marcosbc At the time we were running 2.3.6, I believe.  I have recently updated to 2.3.7.  I basically just blew away what we had and started over.

Given your comments. I believe you are correct.  We had patterned the kubernetes volumes after a helm chart we had found online which suggested mounting `/bitnami`.  I'll back up the site and bounce the postgresql database to see if that replicates the issue.  Then I will update the volume mapping and repeat. I can report back here.
